### PR TITLE
Topological sort, zero_grads

### PR DESCRIPTION
**Pull request by adamritter**

### CODE REVIEW

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -108,9 +108,27 @@ def randn(*shape):
...
+      for i in self._ctx.parents:
+        if i not in visited:
+          i.deepwalk(visited, a)
+      a.append(self)
+    return a
+
   def backward(self, allow_fill=True):
     #print("running backward on", self)
-    if self._ctx is None:
+    if (self._ctx is None):
```
Comment (geohot) [inline tinygrad/tensor.py:131]: Why change this?

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -108,9 +108,27 @@ def randn(*shape):
   def eye(dim):
     return Tensor(np.eye(dim).astype(np.float32))
 
+  def zero_grads(self):
```
Comment (geohot) [inline tinygrad/tensor.py:111]: This doesn't belong here like this. Put in the optimizer?

File: tinygrad/tensor.py [hunk 3]
```diff
@@ -108,9 +108,27 @@ def randn(*shape):
...
+    if self.grad is not None:
+      self.grad = None
+      if self._ctx:
+        for parent in self._ctx.parents:
+          parent.zero_grads()
+
+  def deepwalk(self, visited=None, a=None):
```
Comment (geohot) [inline tinygrad/tensor.py:118]: Put this in the backward function?

Comment (geohot) [inline tinygrad/tensor.py:118]: Put this in the backward function, check out how micrograd does it:

https://github.com/karpathy/micrograd/blob/master/micrograd/engine.py#L54

File: tinygrad/tensor.py [hunk 4]
```diff
@@ -119,21 +137,23 @@ def backward(self, allow_fill=True):
...
+      if len(t0._ctx.parents) == 1:
+        grads = [grads]
+      for t,g in zip(t0._ctx.parents, grads):
+        if g is None:
+          continue
+        if g.shape != t.data.shape:
+          print("grad shape must match tensor shape in %r, %r != %r" %
+            (self._ctx, g.shape, t.data.shape))
+          assert(False)
+        if t.grad is None:
```
Comment (geohot) [inline tinygrad/tensor.py:153]: t.grad = Tensor(g) if t.grad is None else (Tensor(g) + t.grad)

File: tinygrad/tensor.py [hunk 5]
```diff
@@ -119,21 +126,30 @@ def backward(self, allow_fill=True):
...
-      if g.shape != t.data.shape:
-        print("grad shape must match tensor shape in %r, %r != %r" %
-          (self._ctx, g.shape, t.data.shape))
-        assert(False)
-      t.grad = Tensor(g)
-      t.backward(False)
+    visited, nodes = set(), []
+    def deepwalk(node):
+      visited.add(self)
+      if node._ctx:
```
Comment (geohot) [inline tinygrad/tensor.py:132]: is this ever not true?

Comment (adamritter) [reply tinygrad/tensor.py:125]: Yes, for all leaf nodes, we don't want to walk them

File: tinygrad/tensor.py [hunk 6]
```diff
@@ -109,7 +109,7 @@ def eye(dim):
     return Tensor(np.eye(dim).astype(np.float32))
 
   def backward(self, allow_fill=True):
-    #print("running backward on", self)
+    # print("running backward on", self)
```
Comment (geohot) [inline tinygrad/tensor.py:112]: Can just delete this print

**Comment by geohot** [comment]
Think all that remains is fixing the autogrid test by clearing the gradient.

**Comment by adamritter** [comment]
> Think all that remains is fixing the autograd test by clearing the gradient.

I don't yet know how to do it, this didn't work:
  for o in range(jo):
    input.grad, output.grad = None, None

FAIL: test_gradcheck (__main__.TestTinygrad)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\Razer\Documents\GitHub\tinygrad\test\test_tensor.py", line 62, in test_gradcheck
    self.assertTrue(gradcheck(tiny_func, tiny_x))
AssertionError: False is not true

======================================================================
FAIL: test_jacobian (__main__.TestTinygrad)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\Razer\Documents\GitHub\tinygrad\test\test_tensor.py", line 51, in test_jacobian
    np.testing.assert_allclose(PJ, J, atol = 1e-5)
  File "c:\users\razer\appdata\local\programs\python\python38\lib\site-packages\numpy\testing\_private\utils.py", line 1532, in assert_allclose
    assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
  File "c:\users\razer\appdata\local\programs\python\python38\lib\site-packages\numpy\testing\_private\utils.py", line 846, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Not equal to tolerance rtol=1e-07, atol=1e-05

Mismatched elements: 40 / 50 (80%)
Max absolute difference: 2.81465007
Max relative difference: 1708.75129443
 x: array([[-0.245939, -0.436551, -0.433157, -0.449366, -0.328601, -0.399411,
        -0.500921, -0.442804, -0.231163, -0.521697],
       [-0.087255, -0.174608,  0.195344,  0.344753, -0.140792,  0.01373 ,...
 y: array([[-2.459392e-01, -4.365509e-01, -4.331574e-01, -4.493659e-01,
        -3.286008e-01, -3.994109e-01, -5.009208e-01, -4.428038e-01,
        -2.311633e-01, -5.216969e-01],...

**Comment by adamritter** [comment]
OK, I fixed it, though it's ugly, because now the passed function is called in the outer and the inner loop as well

**Pull request merged**